sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java

```diff
@@ -0,0 +1,448 @@
+                            "Unsupported floating-point precision: " + type.getPrecision().name());
+                    }
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Utf8 type) {
+                    return FieldType.STRING;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Binary type) {
+                    return FieldType.BYTES;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.FixedSizeBinary type) {
+                    return FieldType.logicalType(FixedBytes.of(type.getByteWidth()));
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Bool type) {
+                    return FieldType.BOOLEAN;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Decimal type) {
+                    // FieldType.DECIMAL isn't perfect here since arrow decimal has a
+                    // scale/precision fixed by the schema, but FieldType.DECIMAL uses a BigDecimal,
+                    // whose precision/scale can change from row to row.
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Date type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Time type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Timestamp type) {
+                    if (type.getUnit() == TimeUnit.MILLISECOND
+                        || type.getUnit() == TimeUnit.MICROSECOND) {
+                      return FieldType.DATETIME;
+                    } else {
+                      throw new IllegalArgumentException(
+                          "Unsupported timestamp unit: " + type.getUnit().name());
+                    }
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Interval type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Duration type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+                });
+    return fieldType.withNullable(arrowFieldType.isNullable());
```

Comment:
Arrow seems to have isNullable on both the field and the field type. Are they guaranteed to both be set?

Reply:
I think FieldType here is just a construction for the java library combining Field and Type. In the flatbuffers metadata nullability is only specified on the [Field](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L302)